File: udpreceiver_stop_wait.py
import socket
import struct
import hashlib
import time
import math

def receive_file():
    # Create a UDP socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    ack_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    # Increase socket receive buffer to 1MB
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 1024 * 1024)
    print('starting up on %s port %s' % ('localhost', 4445))

    server_address = ('localhost', 4444)
    server_ack_adress = ('localhost', 4447)
    sock.bind(server_address)

    md5 = hashlib.md5()

    # Initialize variables
    transmission_id = None
    max_seq_number = None
    file_name = None
    file_data = b''
    received_md5 = None

    bytes_received = 0 

    print("listening on %s:%s" % server_address)

    while True:
        data, _ = sock.recvfrom(1024*64)

        #print(f"received {len(data)} bytes")
        bytes_received += len(data)
        
        if not transmission_id:
            # This is the first packet, get transmissionId, maxSeqNumber, and file name
            start_time = int(time.time() * 1000)
            transmission_id, seq_number, max_seq_number = struct.unpack('!iii', data[:12])
            file_name = data[12:].decode('utf-8', 'ignore')
            print('Receiving file:', file_name)
            print('seq_number:', seq_number)
            print('max_seq_number:', max_seq_number)
        else:
            # Regular packet, get data
            seq_number = struct.unpack('!i', data[4:8])[0]
            if seq_number == max_seq_number:
                print(f"final: {seq_number}")
                # get md5
                received_md5 = data[8:]
                break
            file_data += data[8:]
            md5.update(data[8:])
            print(f"\r{seq_number}/{max_seq_number}", end='')
        # send ACK with seq_number
        #time.sleep(0.01)
        ack = struct.pack('!ii', 2, seq_number) # TODO: Fix rec receiving it's own ack
        ack_sock.sendto(ack, server_ack_adress)

    # Check MD5 hash
    calculated_md5 = md5.digest()

    current_time_ms = int(time.time() * 1000)
    receive_time = current_time_ms - start_time
    print(f"Time taken: {receive_time} ms")

    if calculated_md5 == received_md5:
        print('File received successfully')
        with open(file_name, 'wb') as f:
            f.write(file_data)
    else:
        print('File was not received correctly')

    sock.close()
    return receive_time, bytes_received

def store_result_in_txt(speed, size):
    size_in_mb = size / 1e6
    with open("results.txt", "a") as f:
        f.write(f"Avg. MBps for {math.floor(size_in_mb)}: {speed}\n")

times_ms = []
sizes_bytes = []
files_sent = 0

while True:
    files_sent += 1
    rcv_time, file_size = receive_file()
    times_ms.append(rcv_time)
    sizes_bytes.append(file_size)
    mbps = (sum(sizes_bytes) / 1e6) / (sum(times_ms) / 1000)
    print(f"Average MBps: {mbps} over {files_sent} files")
    if files_sent == 10:
        store_result_in_txt(mbps, sizes_bytes[0])
        times_ms = []
        sizes_bytes = []
        files_sent = 0

File: udp/src/UDPReceiverStopWait.java
package udp.src;

import java.io.*;
import java.net.*;
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

public class UDPReceiverStopWait {

    private static final int BUFFER_SIZE = 64 * 1024; // 64KB

    public static long receiveFile() throws IOException, NoSuchAlgorithmException {
        DatagramSocket socket = new DatagramSocket(4444);
        DatagramSocket ackSocket = new DatagramSocket();
        socket.setReceiveBufferSize(1024 * 1024); // Set receive buffer to 1MB
        InetAddress address = InetAddress.getByName("localhost");
        int ackPort = 4447;

        System.out.println("starting up on localhost port 4445");

        MessageDigest md5 = MessageDigest.getInstance("MD5");

        // Initialize variables
        Integer transmissionId = null;
        Integer maxSeqNumber = null;
        String fileName = null;
        ByteArrayOutputStream fileData = new ByteArrayOutputStream();
        byte[] receivedMd5 = null;

        System.out.println("listening on localhost:4444");

        long startTime = 0;

        while (true) {
            byte[] buffer = new byte[BUFFER_SIZE];
            DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
            socket.receive(packet);

            int packetLength = packet.getLength();
            //System.out.println("received " + packetLength + " bytes");

            ByteBuffer wrapped = ByteBuffer.wrap(packet.getData(), 0, packetLength);

            if (transmissionId == null) {
                // This is the first packet, get transmissionId, maxSeqNumber, and file name
                startTime = System.currentTimeMillis();
                transmissionId = wrapped.getInt();
                int seqNumber = wrapped.getInt();
                maxSeqNumber = wrapped.getInt();
                byte[] fileNameBytes = new byte[packetLength - 12];
                wrapped.get(fileNameBytes);
                fileName = new String(fileNameBytes).trim();
                System.out.println("Receiving file: " + fileName);
                System.out.println("seq_number: " + seqNumber);
                System.out.println("max_seq_number: " + maxSeqNumber);
            } else {
                // Regular packet, get data
                wrapped.position(4); // skip transmissionId
                int seqNumber = wrapped.getInt();

                if (seqNumber == maxSeqNumber) {
                    System.out.println("final: " + seqNumber);
                    // get md5
                    receivedMd5 = new byte[packetLength - 8];
                    wrapped.get(receivedMd5);
                    break;
                }

                byte[] data = new byte[packetLength - 8];
                wrapped.get(data);
                fileData.write(data);
                md5.update(data);
                System.out.print("\rseq_number: " + seqNumber);
                System.out.flush();
            }

            // Send ACK with seq_number
            ByteBuffer ackBuffer = ByteBuffer.allocate(8);
            ackBuffer.putInt(2);
            ackBuffer.putInt(wrapped.getInt(4));
            byte[] ack = ackBuffer.array();
            DatagramPacket ackPacket = new DatagramPacket(ack, ack.length, address, ackPort);
            ackSocket.send(ackPacket);
        }

        // Check MD5 hash
        byte[] calculatedMd5 = md5.digest();

        long receiveTime = System.currentTimeMillis() - startTime;
        System.out.println("Time taken: " + receiveTime + " ms");
        System.out.println("Speed: " + (fileData.size() / 1e6) / (receiveTime / 1000.0) + " MBps");

        if (MessageDigest.isEqual(calculatedMd5, receivedMd5)) {
            System.out.println("File received successfully");
            try (FileOutputStream fos = new FileOutputStream(fileName)) {
                fileData.writeTo(fos);
            }
        } else {
            System.out.println("File was not received correctly");
        }

        socket.close();
        return receiveTime;
    }

    public static void main(String[] args) throws IOException, NoSuchAlgorithmException {
        List<Long> times = new ArrayList<>();
        int filesSent = 0;

        while (true) {
            filesSent++;
            long rcvTime = receiveFile();
            times.add(rcvTime);
            System.out.println("Average time: " + times.stream().mapToLong(Long::longValue).average().orElse(0.0) + " ms over " + filesSent + " files");
        }
    }
}
